src/go/k8s/apis/core/v1alpha1/redpandacluster_types.go

```diff
@@ -0,0 +1,94 @@
+// Copyright 2021 Vectorized, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+// Package v1alpha1 represent Custom Resource definition of the vectorized.io core group
+package v1alpha1
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// RedpandaClusterSpec defines the desired state of RedpandaCluster
+type RedpandaClusterSpec struct {
+	// Important: Run "make" to regenerate code after modifying this file
+
+	// Version of Redpanda to use from upstream.
+	Version	string	`json:"version,omitempty"`
+	// Replicas determine how big the cluster will be.
+	// +kubebuilder:validation:Minimum=0
+	Replicas	*int32	`json:"replicas,omitempty"`
+	// Resources 1 Redpanda container will use.
+	// To calculate overall resource consumption one need to
+	// multiply replicas against limits
+	Resources	corev1.ResourceRequirements	`json:"resources"`
+	// Configuration represent redpanda specific configuration
+	Configuration	RedpandaConfig	`json:"configuration,omitempty"`
+}
+
+// RedpandaClusterStatus defines the observed state of RedpandaCluster
+type RedpandaClusterStatus struct {
+	// Important: Run "make" to regenerate code after modifying this file
+
+	// Replicas show how many nodes are working in the cluster
+	// +optional
+	Replicas	int32	`json:"replicas,omitempty"`
+	// Nodes of the provision redpanda nodes
+	// +optional
+	Nodes	[]string	`json:"nodes,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+
+// RedpandaCluster is the Schema for the redpandaclusters API
+type RedpandaCluster struct {
+	metav1.TypeMeta		`json:",inline"`
+	metav1.ObjectMeta	`json:"metadata,omitempty"`
+
+	Spec	RedpandaClusterSpec	`json:"spec,omitempty"`
+	Status	RedpandaClusterStatus	`json:"status,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+
+// RedpandaClusterList contains a list of RedpandaCluster
+type RedpandaClusterList struct {
+	metav1.TypeMeta	`json:",inline"`
+	metav1.ListMeta	`json:"metadata,omitempty"`
+	Items		[]RedpandaCluster	`json:"items"`
+}
+
+// RedpandaConfig is the definition of the main configuration
+type RedpandaConfig struct {
```

Comment:
@dotnwat these will be arrays for all TCP+RPC endpoints i think.

Reply:
I would like to start the discussion around `RedpandaConfig` which is defined in [configuration.h](https://github.com/vectorizedio/redpanda/blob/dev/src/v/config/configuration.h#L34-L110). The `rpk` doesn't have complete definition of `redpanda` [configuration](https://github.com/vectorizedio/redpanda/blob/dev/src/go/rpk/pkg/config/definition.go#L24-L35). The operator can not import the `rpk` definition, because `controller-gen` will generate wrong [zz_generated.deepcopy.go](https://github.com/RafalKorepta/redpanda/blob/feature/operator-merge/src/go/k8s/apis/core/v1alpha1/zz_generated.deepcopy.go) code (doesn't compile). The solution to this problem could be:
1.  Create generator that will create all fields in `configuration.h` and `configuration.cc` and `redpanda_conf.go` based on some arbitrary higher level definition (maybe in json/yaml format).
2. Create linter that will validate `rpk`, `operator` and `redpanda` configration files.
3. Create abstraction above the advance `redpanda` configuration which will have different knobs. The operator would be responsible for reconcile desire specification and runtime environment to create `redpanda` advance configuration that will fit.

Reply:
I will leave this generator for the next PR. Current definition is the bare minimum for operator to work.